### ISSUE TITLE
Add filter configuration to log appender

### ISSUE THREAD

**Issue by wallezhang** [open-telemetry/opentelemetry-java-instrumentation#5464]
**Is your feature request related to a problem? Please describe.**
Agent will instrument all log appenders so that all logs will be exported to collector. I just want to export logs on specified appenders.

**Describe the solution you'd like**
I think we can add configuration to filter exporting logs by appender name or log level. Configurations maybe like `otel.instrumentation.log.enabled-appender-names`.

**Comment by rahuldimri** [comment]
Can I work on this request ?

**Comment by mateuszrzeszutek** [comment]
Let me fill in some of the requirements here.

I think we could introduce a set of new configuration properties for this:
* For javaagent instrumentations, something like `otel.instrumentation.<instrumentation-name>.experimental.logger-name-pattern`, where `<instrumentation-name>` is one of the [instrumentation names](https://opentelemetry.io/docs/instrumentation/java/automatic/agent-config/#suppressing-specific-agent-instrumentation). This will make it similar to the other logging instrumentation settings.
* For library instrumentations, a new configuration setting on appender/builder.
* The value of this property would need to be a valid `java.util.regex.Pattern`.
* If this property is not set log events are not filtered at all.
* If this property is set, log events that do not match this pattern will not emit any telemetry.

The following instrumentations need to be updated:
* JUL
* JBoss LogManager
* Log4j 1
* Log4j 2, both javaagent and library instrumentations
* Logback, both javaagent and library instrumentations

**Comment by rahuldimri** [comment]
Yes let me give a try !!

**Comment by trask** [comment]
I was assuming this issue was about "appender name", so that you could configure one appender with the particular configuration that you want, and then ask the javaagent to pick up only logs that flow through that one appender?

**Comment by mateuszrzeszutek** [comment]
> I was assuming this issue was about "appender name", so that you could configure one appender with the particular configuration that you want, and then ask the javaagent to pick up only logs that flow through that one appender?

Indeed, I somehow missed that and got fixated on logger name instead 🙈 

In that case, this feature will only concern javaagent logging instrumentations; when you're using library appenders you can already configure them however you want.

How about a list-type configuration property named`otel.instrumentation.<instrumentation-name>.experimental.include-appenders` (or perhaps `...exclude-appenders` to match several `...exclude-methods` configs we already have)?

**Comment by wallezhang** [comment]
> > I was assuming this issue was about "appender name", so that you could configure one appender with the particular configuration that you want, and then ask the javaagent to pick up only logs that flow through that one appender?
> 
> Indeed, I somehow missed that and got fixated on logger name instead 🙈
> 
> In that case, this feature will only concern javaagent logging instrumentations; when you're using library appenders you can already configure them however you want.
> 
> How about a list-type configuration property named`otel.instrumentation.<instrumentation-name>.experimental.include-appenders` (or perhaps `...exclude-appenders` to match several `...exclude-methods` configs we already have)?

Can we use the same configuration in javaagent instrumentation?

**Comment by JessHolle** [comment]
So is the thought here to instead automatically add/apply OpenTelemetryAppender wherever specific other appenders are applied?  Or to instrument those appenders code and check appender name rather than instrumenting loggers?

**Comment by trask** [comment]
> Or to instrument those appenders code and check appender name rather than instrumenting loggers?

I'm thinking something like this could be nice.

Then as a configuration property, you could set the appender name that you want the javaagent to capture from, and if that configuration property is set, the javaagent could suppress the general instrumentation and only instrument that specific appender.

**Comment by JessHolle** [comment]
> > Or to instrument those appenders code and check appender name rather than instrumenting loggers?
> 
> I'm thinking something like this could be nice.
> 
> Then as a configuration property, you could set the appender name that you want the javaagent to capture from, and if that configuration property is set, the javaagent could suppress the general instrumentation and only instrument that specific appender.

I'd think a set of appender names would be ideal -- a Set.contains() check should be cheap enough and allow maximum flexibility, e.g. to capture all of the data the original app routed to several different log files.

**Comment by xiangtianyu** [comment]
Is there any progress on this issue recently?

**Comment by mateuszrzeszutek** [comment]
Hey @xiangtianyu ,
No, no changes.

**Comment by xiangtianyu** [comment]
> Hey @xiangtianyu , No, no changes.

Is there any plan for this? Or is there any discussion about this?

**Comment by mateuszrzeszutek** [comment]
> Is there any plan for this? Or is there any discussion about this?

I believe none of the maintainers of this project have cycles to handle that.

**Comment by trask** [comment]
I'd suggest having the configuration be a "map", where key is logger name/prefix and value is min level to capture for that logger name/prefix

**Comment by wallezhang** [comment]
> I'd suggest having the configuration be a "map", where key is logger name/prefix and value is min level to capture for that logger name/prefix

I want to try this issue. Please assign to me.

**Comment by trask** [comment]
This is being worked on at the specification level now through a combination of
* https://github.com/open-telemetry/opentelemetry-configuration/pull/140
* https://github.com/open-telemetry/opentelemetry-specification/pull/4290

(and some followups will be needed still)
